src/mlpack/methods/ann/loss_functions/multilabel_softmargin_loss_impl.hpp

```diff
@@ -0,0 +1,79 @@
+/**
+ * @file multilabel_softmargin_loss_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the Multi Label Soft Margin Loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_IMPL_HPP
+
+// In case it hasn't been included.
+#include "multilabel_softmargin_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MultiLabelSoftMarginLoss<InputDataType, OutputDataType>::
+MultiLabelSoftMarginLoss(
+    arma::mat weight,
+    const size_t num_classes,
+    const bool reduction) :
+    num_classes(num_classes),
+    reduction(reduction)
+{
+  class_weights.ones(1, num_classes);
+  class_weights = weight;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+typename InputType::elem_type
+MultiLabelSoftMarginLoss<InputDataType, OutputDataType>::Forward(
+    const InputType& input, const TargetType& target)
+{
+  InputType logSigmoid = arma::log((1 / (1 + arma::exp(-input))));
+  InputType logSigmoidNeg = arma::log(1 / (1 + arma::exp(input)));
+  InputType loss = arma::mean(arma::mean(-(target % logSigmoid +
+      (1 - target) % logSigmoidNeg), 1) * class_weights, 1);
+
+  if (reduction)
+    return arma::as_scalar(arma::sum(loss));
+
+  return arma::as_scalar(arma::mean(loss));
```

Comment:
Hi @iamshnoo, Could you help me in understanding why we are taking the mean twice in line 42. Since weights might not be needed. I think this will be handled once you handle the above comments.

Reply:
Assume a 3*3 input (as in the test case).
The inner mean computes a 3*3 matrix with 3 identical columns because of the multiplication with the weights matrix. To go from 3 columns to 1 column (so that the output matches the form of output in the PyTorch implementation with reduction='none'), I use the outer mean.

I have explained the need for the weights parameter in the comment above. Let's continue this discussion in that thread for simplicity.